scripts/rpmbuildinfo

```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+
+#
+# Copyright (c) 2016  Bernhard M. Wiedemann <bernhard@zq1.de>
+# Copyright (c) 2016  Dennis Gilmore <dennis@ausil.us>
+# Copyright (c) 2016  Simon "HW42" Geiser <hw42@ipsumj.de>
+# Copyright (c) 2016  Wojciech Porczyk <woju@invisiblethingslab.com>
+# Copyright (c) 2019  Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>
+# Copyright (c) 2021  Frédéric Pierret <frederic.pierret@qubes-os.org>
+#
+# All rights reserved.
+# 
+# Permission to use, copy, modify, and distribute this software for any purpose
+# with or without fee is hereby granted, provided that the above copyright
+# notice and this permission notice appear in all copies.
+# 
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT OF THIRD PARTY RIGHTS. IN
+# NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+# OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
+# OR OTHER DEALINGS IN THE SOFTWARE.
+# 
+# Except as contained in this notice, the name of a copyright holder shall not
+# be used in advertising or otherwise to promote the sale, use or other dealings
+# in this Software without prior written authorization of the copyright holder.
+#
+
+set -e -o pipefail
+
+getos() {
+    # shellcheck disable=SC1091
+    test -r /etc/os-release && . /etc/os-release
+    if test -z "${ID}"; then
+        ID="$(cat /etc/system-release)"
+    fi
+    printf '%s' "${ID}"
+}
+
+RPM_BUILD_ROOT="$1"
+RPM_BUILD_DIR="$2"
```

Comment:
BUILDROOT and BUILD_DIR may leak relatively sensitive information about usernames etc and are a slap in the face of reproducability for no good reason at all. If anything, inclusion of these should be an opt-in feature (eg enable for troubleshooting purposes only)

Reply:
Unfortunately, this is necessary because it can affect the build significantly. Those paths regularly get embedded in source files, debuginfo, and documentation metadata. That's why I had to change redhat-rpm-config last year to stop Koji builds with doxygen doc builds from failing.

Reply:
> Unfortunately, this is necessary because it can affect the build significantly. Those paths regularly get embedded in source files, debuginfo, and documentation metadata. That's why I had to change redhat-rpm-config last year to stop Koji builds with doxygen doc builds from failing.

That embedding is a bug in the package in question and should be reported as such.  BUILDROOT and BUILD_DIR should be used only when bugs require it.

Reply:
Rpm checks against embedded BUILDROOT in the packages already and fails the build when found, we should probably extend that to cover BUILD DIR because references to that are equally a bug, always. Had to look up the redhat-rpm-config change - Doxygen needs a whack in the head, really. I also believe it has options to disable that, if not then we should file a bug on it.

Yeah the "security" aspect is lame, and source + debuginfo packages having them is one thing, it's just that for a moment I thought we're talking about the main binary packages ... but we're not, this is supposed to go into a subpackage. Which I'm not sure makes a whole lot of sense in the first place: you have to fish the plaintext file from inside an rpm. It'd be different if the relevant data could be queried in an rpm header itself, but we don't want to add two dozen new tags specific for just this either. Is there a rationale for this particular design somewhere? (it should be in the commit messages of this thing!)

Reply:
> > Unfortunately, this is necessary because it can affect the build significantly. Those paths regularly get embedded in source files, debuginfo, and documentation metadata. That's why I had to change redhat-rpm-config last year to stop Koji builds with doxygen doc builds from failing.
> 
> That embedding is a bug in the package in question and should be reported as such. BUILDROOT and BUILD_DIR should be used only when bugs require it.

Sorry, I don't care whether it's a bug or not. It's ***reality***. Since that's a required component of build reproducibility, it needs to be recorded.

Reply:
> Rpm checks against embedded BUILDROOT in the packages already and fails the build when found, we should probably extend that to cover BUILD DIR because references to that are equally a bug, always.

Where is this check?  It wasn’t obvious from a quick search.

>  (it should be in the commit messages of this thing!)

@fepitre time to add them!

Reply:
> > > Unfortunately, this is necessary because it can affect the build significantly. Those paths regularly get embedded in source files, debuginfo, and documentation metadata. That's why I had to change redhat-rpm-config last year to stop Koji builds with doxygen doc builds from failing.
> > 
> > 
> > That embedding is a bug in the package in question and should be reported as such. BUILDROOT and BUILD_DIR should be used only when bugs require it.
> 
> Sorry, I don't care whether it's a bug or not. It's _**reality**_. Since that's a required component of build reproducibility, it needs to be recorded.

What if RPM uses Linux namespaces to set fixed values for these?

Reply:
Then I want to know what those values are anyway, because there's no guarantee that another environment is going to use the same fixed values without knowing what those are.

Reply:
> Where is this check? It wasn’t obvious from a quick search.

It's one of those buildroot  policy scripts, see https://github.com/rpm-software-management/rpm/blob/master/scripts/check-buildroot

Reply:
> Is there a rationale for this particular design somewhere?

It was discussed in the thread mentioned in the PR description: https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/OLCD2L6I5KUKWR6WLIUXFUYZ7KAQ66E3/

Reply:
Uh, right. So how about we discuss the design *here* instead.

I'm not totally up to speed with the intended purposes of this file, but assuming collecting this info is useful for one or more purposes (in principle I can agree with that)... One of the purposes seems to be allowing comparison between builds, but for that, placing the info inside sub-packages which you have to install (or at least unpack first) to even look at seems painful to say the least.

Sticking it into an rpm just because it's generated by rpm... I'm not sure it makes sense here. And even if it did, I don't know why anybody would want to *install* such a thing on their system (feel free to enlighten me). Since we're talking about a simple plain-text file here, it could just as well be stuffed into the header of an src.rpm-like package so you can compare packages by just querying them.

Reply:
See [here](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/OLCD2L6I5KUKWR6WLIUXFUYZ7KAQ66E3/#7QOKE7LUJNG6VJWENGVF7TFEOM2MX7JU) why not src.rpm itself. You don't need to install rpm to get the file, you just need to extract its payload (rpmdev-extract, or any of the other tools).

Why rpm instead of just plain file? Because we've got all the infrastructure and tooling for handling rpms already (Debian decided to ship it as a separate plain file, and _still_ hasn't figured out final solution for distributing them... https://bugs.debian.org/763822).

My concern about putting the buildinfo file (either as a file, or a header tag) into the main binary output itself is possibly harming reproducibility - for example if a package builds reproducible regardless of the host arch, yet the buildinfo includes host arch, then the output won't be reproducible anymore (when you try to reproduce on another host arch). Ideally the buildinfo should not include stuff that isn't relevant for reproducing the specific package, but that set can vary between packages, and IMHO it isn't realistic to maintain this info on per-package basis.

Reply:
Note that I suggested putting it into "an src.rpm like package", which would be a separate (sub-)package, but not necessarily a regular *binary* package which is expected to be installed, nor the main src.rpm (which would be the right place from rpm POV in many ways but not so much for the practical world). The main package is indeed the wrong place for it for various reasons.

Reply:
If we want to stick it somewhere, the debugsource package might be as good of a place as any?

Reply:
> If we want to stick it somewhere, the debugsource package might be as good of a place as any?

Do all packages have debugsource? AFAIK noarch usually do not, right?

Reply:
Oh that's true...

Reply:
A separate buildinfo RPM seems like the best option to me.